### PR TITLE
TILA-2299 | ReservationUnit published filter and reservation validation adjustments

### PR DESCRIPTION
**Pull request by nicobav**
## Change log
  - Adjust ReservationUnit published gql filter so that it checks between publish ends and begins
  - Scheduler logic change for determining reservation period end date
     - If the ReservatioUnit's reservation_ends is after the reservation start period it uses the the reservation_end date
  - Raise validation error in reservation creation if reservation unit is draft or archived.

Refs TILA-2299

### CODE REVIEW

File: api/graphql/reservation_units/reservation_unit_filtersets.py [hunk 1]
```diff
@@ -199,7 +199,11 @@ def get_is_visible(self, qs, property, value):
         today = datetime.datetime.now(tz=get_default_timezone())
```
[the remark this answers is not in the document]
Comment (nicobav) [reply api/graphql/reservation_units/reservation_unit_filtersets.py:199]: Your point is good, I'll change this nevertheless it isn't part of the scope (but while I'm changing this to the other line as well) 👍

File: api/graphql/reservations/reservation_serializers/mixins.py [hunk 2]
```diff
@@ -183,18 +184,56 @@ def calculate_price(
...
+        )
+
+        reservation_in_non_published_reservation_unit = (
+            reservation_unit.publish_ends
+            and now >= reservation_unit.publish_ends
+            and (
+                reservation_unit.publish_begins is None
+                or (reservation_unit.publish_begins <= reservation_unit.publish_ends)
+            )
+        )
```
[the remark this answers is not in the document]
Comment (nicobav) [reply api/graphql/reservations/reservation_serializers/mixins.py:215]: It should be ok since it means that its published.

@vergama can you confirm?

[the remark this answers is not in the document]
Comment (tuomas777) [reply api/graphql/reservations/reservation_serializers/mixins.py:216]: Ah ok, for some reason I thought it isn't published when `publish_begins` is `None`. All good then 👍

File: api/graphql/reservation_units/reservation_unit_filtersets.py [hunk 3]
```diff
@@ -196,10 +195,14 @@ def get_min_persons_lte(self, qs, property, value):
...
+        now = timezone.now()
         qs = qs.filter(is_draft=False, is_archived=False)
-        published = (Q(publish_begins__lte=today) | Q(publish_begins__isnull=True)) & (
-            Q(publish_ends__gt=today) | Q(publish_ends__isnull=True)
+        published = (Q(publish_begins__lte=now) | Q(publish_begins__isnull=True)) & (
+            Q(publish_ends__gt=now)
+            | Q(publish_ends__isnull=True)
+            | Q(
+                publish_ends__lt=F("publish_begins")
+            )  # When end is before begin and begin is today or before it is considered as (re)published.
```
Comment (tuomas777) [inline api/graphql/reservation_units/reservation_unit_filtersets.py:205]: I think this comment is now slightly inaccurate because if begin is later today the unit is not considered as republished right now.

File: reservation_units/utils/reservation_unit_reservation_scheduler.py [hunk 4]
```diff
@@ -136,8 +136,11 @@ def _get_reservation_period_start(self) -> datetime.datetime:
...
         return DEFAULT_TIMEZONE.localize(datetime.datetime.now() + delta)
 
-    def _get_reservation_period_end(self, start: datetime.date) -> datetime.date:
-        if self.reservation_unit.reservation_ends:
+    def _get_reservation_period_end(self, start: datetime.datetime) -> datetime.date:
+        if (
+            self.reservation_unit.reservation_ends
+            and self.reservation_unit.reservation_ends > start
+        ):
             return self.reservation_unit.reservation_ends.date()
```
Comment (tuomas777) [inline reservation_units/utils/reservation_unit_reservation_scheduler.py:144]: Again not part of this PR, and I'm not actually sure about this, but I think this might give the date based on UTC timezone, so it will be incorrect when the time is close to midnight. That is probably not very common in practice though 😄

**Review by tuomas777 — APPROVED** [review]
Added a couple of minor comments you can check and ignore if they are incorrect 😄 

Overall looks very good to me 👍 🚀 🐳
